python_modules/dagster/dagster/_core/definitions/asset_graph.py

```diff
@@ -335,6 +350,75 @@ def toposort_asset_keys(self) -> Sequence[AbstractSet[AssetKey]]:
     def has_self_dependency(self, asset_key: AssetKey) -> bool:
         return asset_key in self.get_parents(asset_key)
 
+    def bfs_filter_subsets(
+        self,
+        condition_fn: Callable[[AssetKey, Optional[PartitionsSubset]], bool],
+        initial_subset: "AssetGraphSubset",
+    ) -> "AssetGraphSubset":
+        """
+        Returns asset partitions within the graph that
+        - Are >= initial_asset_partitions
+        - Asset matches the condition_fn
+        - Any of their ancestors >= initial_asset_partitions match the condition_fn
+
+        Visits parents before children.
+        """
+        from .asset_graph_subset import AssetGraphSubset
+
+        all_assets = set(initial_subset.asset_keys)
+        check.invariant(
+            len(initial_subset.asset_keys) == 1, "Multiple initial assets not yet supported"
+        )
+        initial_asset_key = next(iter(initial_subset.asset_keys))
+        queue = deque(
+            [
+                (
+                    initial_asset_key,
+                    initial_subset.get_partitions_subset(initial_asset_key)
+                    if self.get_partitions_def(initial_asset_key)
+                    else None,
+                )
+            ]
+        )
+        result = AssetGraphSubset(self)
+
+        while len(queue) > 0:
+            asset_key, partitions_subset = queue.popleft()
+
+            if condition_fn(asset_key, partitions_subset):
+                result |= AssetGraphSubset(
+                    self,
+                    non_partitioned_asset_keys={asset_key} if partitions_subset is None else set(),
+                    partitions_subsets_by_asset_key={asset_key: partitions_subset}
+                    if partitions_subset is not None
+                    else {},
+                )
+
+                for child in self.get_children(asset_key):
+                    partition_mapping = self.get_partition_mapping(child, asset_key)
+                    child_partitions_def = self.get_partitions_def(child)
+
+                    if child_partitions_def:
+                        if partitions_subset is None:
+                            child_partitions_subset = (
+                                child_partitions_def.subset_with_all_partitions()
+                            )
+                        else:
+                            child_partitions_subset = (
+                                partition_mapping.get_downstream_partitions_for_partitions(
+                                    partitions_subset,
+                                    downstream_partitions_def=child_partitions_def,
+                                )
+                            )
+                    else:
+                        child_partitions_subset = None
+
+                    if child not in all_assets:
+                        queue.append((child, child_partitions_subset))
+                        all_assets.add(child)
+
+        return result
```

Comment:
This looks good to me, but I think it'd be possible to compose a graph that fails this logic with something whacky like this:

```
A -> C (last 7 days) -> D
A -> B -> C (last 14 days) -> D
```

(Since it would choose only the last 7 days of C)

Reply:
@bengotow would you mind explaining your example in a little more detail? I'm having some trouble pinning down the notation.

Reply:
Hey @sryza! Ahh sorry, in my example there are two assets in the graph that depend on different partitions of the same asset "C". The one that is visited first by this code is the one with the shortest path distance since this is a breadth-first-search.  We see A->C and put "child_partitions_subset = last 7 days" into the `queue` array on line 435. We never visit the B->C edge because `C` is already in `all_assets`, which means we don't visit the "last 14 days" dependency.

In general this algorithm does not support the same asset being referenced multiple times in the graph with different partition subsets. That might be totally fine, was mostly just calling out that it would generate an incorrect list of child partitions.

Reply:
Ah gotcha - yeah, good point. Probably not super common, but I updated the PR to fix this for the sake of completeness